src/cursor/cur_hs.c

```diff
@@ -658,6 +658,87 @@ __curhs_insert(WT_CURSOR *cursor)
     API_END_RET(session, ret);
 }
 
+/*
+ * __curhs_update --
+ *     WT_CURSOR->update method for the hs cursor type.
+ */
+static int
+__curhs_update(WT_CURSOR *cursor)
+{
+    WT_CURSOR *file_cursor;
+    WT_CURSOR_BTREE *cbt;
+    WT_CURSOR_HS *hs_cursor;
+    WT_DECL_ITEM(hs_value);
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    WT_UPDATE *hs_tombstone, *hs_upd;
+
+    uint64_t hs_upd_type;
+    wt_timestamp_t hs_durable_ts, hs_stop_durable_ts;
+
+    hs_cursor = (WT_CURSOR_HS *)cursor;
+    file_cursor = hs_cursor->file_cursor;
+    cbt = (WT_CURSOR_BTREE *)file_cursor;
+    hs_tombstone = hs_upd = NULL;
+
+    CURSOR_API_CALL_PREPARE_ALLOWED(cursor, session, update, CUR2BT(file_cursor));
+
+    /* We are assuming that the caller has already searched and found the key. */
+    WT_ASSERT(session, F_ISSET(file_cursor, WT_CURSTD_KEY_SET | WT_CURSTD_VALUE_SET));
+
+    /*
+     * Only valid scenario to update the history store is to add the stop timestamp. Any other case
+     * should fail.
+     */
+    WT_ASSERT(session, !WT_TIME_WINDOW_IS_EMPTY(&hs_cursor->time_window));
+    WT_ASSERT(session, WT_TIME_WINDOW_HAS_STOP(&hs_cursor->time_window));
+
+    /* We are positioned on the newest value for user key. */
+    WT_ASSERT(session, F_ISSET(hs_cursor, WT_HS_CUR_COUNTER_SET | WT_HS_CUR_TS_SET));
+
+    /* The tombstone to represent the stop time window. */
+    WT_ERR(__wt_upd_alloc_tombstone(session, &hs_tombstone, NULL));
+    hs_tombstone->start_ts = hs_cursor->time_window.stop_ts;
+    hs_tombstone->durable_ts = hs_cursor->time_window.durable_stop_ts;
+    hs_tombstone->txnid = hs_cursor->time_window.stop_txn;
+
+    /* Modify the existing value with a new stop timestamp. */
+
+    /* Allocate a buffer for the history store value. */
+    WT_ERR(__wt_scr_alloc(session, 0, &hs_value));
+
+    /* Retrieve the existing update value and stop timestamp. */
+    WT_ERR(file_cursor->get_value(
+      file_cursor, &hs_stop_durable_ts, &hs_durable_ts, &hs_upd_type, hs_value));
+    WT_ASSERT(session, hs_stop_durable_ts == WT_TS_MAX);
+    WT_ASSERT(session, (uint8_t)hs_upd_type == WT_UPDATE_STANDARD);
+
+    /* Use set_value method to pack the new value. */
+    file_cursor->set_value(
+      file_cursor, hs_cursor->time_window.stop_ts, hs_durable_ts, hs_upd_type, hs_value);
+
+    WT_ERR(__wt_upd_alloc(session, &file_cursor->value, WT_UPDATE_STANDARD, &hs_upd, NULL));
+    hs_upd->start_ts = hs_cursor->time_window.start_ts;
+    hs_upd->durable_ts = hs_cursor->time_window.durable_start_ts;
+    hs_upd->txnid = hs_cursor->time_window.start_txn;
+
+    /* Connect the tombstone to the update. */
+    hs_tombstone->next = hs_upd;
+
+    /* Search the page and insert the updates. */
+    WT_WITH_PAGE_INDEX(session, ret = __wt_hs_row_search(cbt, &file_cursor->key, true));
+    WT_ERR(ret);
+    WT_ERR(__wt_hs_modify(cbt, hs_tombstone));
```

Comment:
I think we need to handle WT_RESTART as well.

Reply:
Fixed.